From 1d19458eec1560ccb400f78369228afc22e614b4 Mon Sep 17 00:00:00 2001
From: Xu Han <hanxu@codeaurora.org>
Date: Mon, 2 Nov 2015 17:38:47 -0800
Subject: [PATCH 29/60] msm: camera: Add NULL check in msm_actuator

NULL check on function pointers and actuator control struct
is necessary to prevent errors.

Change-Id: I212e8558c649e0d432d232dbb99e58b7a064829b
Signed-off-by: Xu Han <hanxu@codeaurora.org>
Signed-off-by: VijayaKumar T M <vtmuni@codeaurora.org>
[haggertk: Port this over to camera_ll]
Signed-off-by: Kevin F. Haggerty <haggertk@lineageos.org>
---
 .../camera_ll/sensor/actuator/msm_actuator.c  | 22 ++++++++++++++++---
 1 file changed, 19 insertions(+), 3 deletions(-)

diff --git a/drivers/media/platform/msm/camera_ll/sensor/actuator/msm_actuator.c b/drivers/media/platform/msm/camera_ll/sensor/actuator/msm_actuator.c
index 7d5c66b1815..41c43b25e5c 100644
--- a/drivers/media/platform/msm/camera_ll/sensor/actuator/msm_actuator.c
+++ b/drivers/media/platform/msm/camera_ll/sensor/actuator/msm_actuator.c
@@ -76,13 +76,23 @@ static int32_t msm_actuator_piezo_set_default_focus(
 static void msm_actuator_parse_i2c_params(struct msm_actuator_ctrl_t *a_ctrl,
 	int16_t next_lens_position, uint32_t hw_params, uint16_t delay)
 {
-	struct msm_actuator_reg_params_t *write_arr = a_ctrl->reg_tbl;
+	struct msm_actuator_reg_params_t *write_arr = NULL;
 	uint32_t hw_dword = hw_params;
 	uint16_t i2c_byte1 = 0, i2c_byte2 = 0;
 	uint16_t value = 0;
-	uint32_t size = a_ctrl->reg_tbl_size, i = 0;
-	struct msm_camera_i2c_reg_array *i2c_tbl = a_ctrl->i2c_reg_tbl;
+	uint32_t size = 0, i = 0;
+	struct msm_camera_i2c_reg_array *i2c_tbl = NULL;
 	CDBG("Enter\n");
+
+	if (a_ctrl == NULL) {
+		pr_err("failed. actuator ctrl is NULL");
+		return;
+	}
+
+	size = a_ctrl->reg_tbl_size;
+	write_arr = a_ctrl->reg_tbl;
+	i2c_tbl = a_ctrl->i2c_reg_tbl;
+
 	for (i = 0; i < size; i++) {
 		/* check that the index into i2c_tbl cannot grow larger that
 		the allocated size of i2c_tbl */
@@ -743,6 +753,12 @@ static int32_t msm_actuator_set_position(
 		return -EFAULT;
 	}
 
+	if (!a_ctrl || !a_ctrl->func_tbl ||
+		!a_ctrl->func_tbl->actuator_parse_i2c_params) {
+		pr_err("failed. NULL actuator pointers.");
+		return -EFAULT;
+	}
+
 	a_ctrl->i2c_tbl_index = 0;
 	for (index = 0; index < set_pos->number_of_steps; index++) {
 		msb = ((set_pos->pos[index] << 7) & 0x80);
-- 
2.17.1

